oracle/oracles/pd.go

```diff
@@ -48,19 +48,108 @@ import (
 	"github.com/tikv/client-go/v2/oracle"
 	pd "github.com/tikv/pd/client"
 	"go.uber.org/zap"
+	"golang.org/x/sync/singleflight"
 )
 
 var _ oracle.Oracle = &pdOracle{}
 
 const slowDist = 30 * time.Millisecond
 
+type adaptiveUpdateTSIntervalState int
+
+const (
+	// adaptiveUpdateTSIntervalStateNormal represents the state that the adaptive update ts interval is synced with the
+	// configuration without performing any automatic adjustment.
+	adaptiveUpdateTSIntervalStateNormal adaptiveUpdateTSIntervalState = iota
+	// adaptiveUpdateTSIntervalStateAdapting represents the state that as there are recently some stale read / snapshot
+	// read operations requesting a short staleness (now - readTS is nearly or exceeds the current update interval),
+	// so that we automatically shrink the update interval. Otherwise, read operations may don't have low resolution ts
+	// that is new enough for checking the legality of the read ts, causing them have to fetch the latest ts from PD,
+	// which is time-consuming.
+	adaptiveUpdateTSIntervalStateAdapting
+	// adaptiveUpdateTSIntervalStateRecovering represents the state that the update ts interval have once been shrunk,
+	// to adapt to reads with short staleness, but there isn't any such read operations for a while, so that we
+	// gradually recover the update interval to the configured value.
+	adaptiveUpdateTSIntervalStateRecovering
+	// adaptiveUpdateTSIntervalStateUnadjustable represents the state that the user has configured a very short update
+	// interval, so that we don't have any space to automatically adjust it.
+	adaptiveUpdateTSIntervalStateUnadjustable
+)
+
+func (s adaptiveUpdateTSIntervalState) String() string {
+	switch s {
+	case adaptiveUpdateTSIntervalStateNormal:
+		return "normal"
+	case adaptiveUpdateTSIntervalStateAdapting:
+		return "adapting"
+	case adaptiveUpdateTSIntervalStateRecovering:
+		return "recovering"
+	case adaptiveUpdateTSIntervalStateUnadjustable:
+		return "unadjustable"
+	default:
+		return "unknown"
+	}
+}
+
+const (
+	// minAllowedAdaptiveUpdateTSInterval is the lower bound of the adaptive update ts interval for avoiding an abnormal
+	// read operation causing the update interval to be too short.
+	minAllowedAdaptiveUpdateTSInterval = 500 * time.Millisecond
```

Comment:
The minimal value of tidb variable `tidb_low_resolution_tso_update_interval` is 10ms, why `minAllowedAdaptiveUpdateTSInterval` is 500ms?

Reply:
To avoid the adaptive update interval mechanism automatically and silently chooses a too low interval can causing too high frequency of getting ts from PD. But we do not reject it if the user configures it explicitly and the low update interval is just what the user expects.